man/man1/any2any.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "ANY2ANY" "1" "December 05, 2021" "any2any 4.0" "User Manual"
.hy
.SH NAME
.PP
any2any - Convert any format media file to any other format media file
.SH SYNOPSIS
.PP
\f[B]any2any\f[R] [ \f[B]-a\f[R] audio codec ] [ \f[B]-v\f[R] video
codec ] [ \f[B]-c\f[R] codec ] [ \f[B]-p\f[R] preset ] [ \f[B]-i\f[R]
input format (lower case) ] [ \f[B]-o\f[R] output format (lower case) ]
[ \f[B]-q\f[R] scale ] [ \f[B]-r\f[R] rate ] [ \f[B]-b\f[R] audio
bitrate ] [ \f[B]-z\f[R] video bitrate ] [ \f[B]-s\f[R] size ] [
\f[B]-t\f[R] threads ] [ \f[B]-I\f[R] ] [ \f[B]-d\f[R] ] [ \f[B]-y\f[R]
] [ \f[B]-u\f[R] ] file1 [file2 ...]
.SH DESCRIPTION
.PP
Converts audio and video media files from one format to another format.
Input and output formats can be specified on the command line.
Convenience links can be created to simplify conversion between formats.
When invoked as the link \f[I]xxx2yyy\f[R], \f[B]any2any\f[R] converts
from media format \f[I]XXX\f[R] to media format \f[I]YYY\f[R].
.PP
The convenience links to \f[I]any2any\f[R] that are installed by default
are:
.IP \[bu] 2
ape2m4a - convert from APE format to M4A
.IP \[bu] 2
ape2mp4 - convert from APE format to MP4
.IP \[bu] 2
avi2mp4 - convert from AVI format to MP4
.IP \[bu] 2
f4v2mp4 - convert from F4V format to MP4
.IP \[bu] 2
flv2mp4 - convert from FLV format to MP4
.IP \[bu] 2
m4a2mp3 - convert from M4A format to MP3
.IP \[bu] 2
m4v2mp4 - convert from M4V format to MP4
.IP \[bu] 2
mkv2mp4 - convert from MKV format to MP4
.IP \[bu] 2
mov2mp4 - convert from MOV format to MP4
.IP \[bu] 2
mp42m4v - convert from MP4 format to M4V
.IP \[bu] 2
mpg2mp4 - convert from MPG format to MP4
.IP \[bu] 2
png2jpg - convert from PNG format to JPG
.IP \[bu] 2
wma2m4a - convert from WMA format to M4A
.IP \[bu] 2
wmv2m4v - convert from WMV format to M4V
.IP \[bu] 2
wmv2mp4 - convert from WMV format to MP4
.PP
Additional convenience links can be created by creating a new symbolic
link to \f[I]any2any\f[R].
For example, to create a convenience link that converts WMA format audio
files to MP3 audio files, issue the following commands:
.IP
.nf
\f[C]
cd /usr/local/bin
sudo ln -s /usr/local/DoctorFreeScript/bin/any2any wma2mp3
\f[R]
.fi
.SH COMMAND LINE OPTIONS
.TP
\f[B]-d\f[R]
indicates tell me what you would do
.TP
\f[B]-i\f[R] input-format
specifies the 3 letter lower case input format
.TP
\f[B]-o\f[R] output-format
specifies the 3 letter lower case output format
.TP
\f[B]-a\f[R] audio-codec
specifies the output audio codec
.TP
\f[B]-v\f[R] video-codec
specifies the output video codec
.TP
\f[B]-s\f[R] size
specifies the output video size (widthxheight)
.TP
\f[B]-c\f[R] codec
specifies the output codec
.TP
\f[B]-I\f[R]
indicates add the converted file to Apple Music
.TP
\f[B]-p\f[R] preset
specifies the ffmpeg preset to use Useful presets: ultrafast superfast
veryfast faster fast medium slow slower veryslow.
Default preset is \[aq]slow\[aq]
.TP
\f[B]-q\f[R] scale
specifies the qscale variable bit rate quality
.TP
\f[B]-r\f[R] rate
specifies the rate for Constant Rate Factor (CRF) encoding.
Use \[dq]-r 0\[dq] to disable for formats other than x264
.TP
\f[B]-b\f[R] bitrate
specifies the bitrate (default 128k)
.TP
\f[B]-t\f[R] threads
specifies the number of threads to use
.TP
\f[B]-y\f[R]
indicates overwrite output files without asking
.TP
\f[B]-u\f[R]
displays this usage message
.PP
Current invocation defaults to:
.PP
\f[V]ffmpeg -i INPUT -loglevel warning -strict -2 -vcodec libx264 -preset slow  -crf 23 -ab 128k  -threads 0 OUTPUT\f[R]
.SH EXAMPLES
.TP
\f[B]any2any WMV/example.mp4\f[R]
The most common use is conversion using the default parameters
.TP
\f[B]any2any -c copy WMV/example.mp4\f[R]
Copy audio and video codecs, convert the container
.TP
\f[B]any2any -s 640x480 WMV/example.mp4\f[R]
Specify output video size of 640 width by 480 height
.TP
\f[B]any2any -v libx264 -p ipod640 WMV/example.mp4\f[R]
Specify libx264 output video codec and use ipod640 preset
.TP
\f[B]any2any -b 512k -I WMV/example.mp4\f[R]
Specify 512k bitrate and add converted file to Apple Music
.TP
\f[B]any2any -v libx264 -r 0 -q 3 WMV/example.mp4\f[R]
Specify libx264 output video codec and variable bit rate
.SH REQUIREMENTS
.PP
This command is basically a wrapper script for \f[B]ffmpeg\f[R].
To install \f[B]ffmpeg\f[R], issue the following command:
.TP
\f[B]Debian Linux\f[R]
sudo apt install ffmpeg
.TP
\f[B]Fedora/CentOS/SuSE Linux\f[R]
sudo yum install ffmpeg
.TP
\f[B]Mac OS X\f[R]
brew install ffmpeg
.SH AUTHORS
.PP
Written by Ronald Record <<github@ronrecord.com>>
.SH LICENSING
.PP
ANY2ANY is distributed under an Open Source license.
See the file LICENSE in the ANY2ANY source distribution for information
on terms & conditions for accessing and otherwise using ANY2ANY and for
a DISCLAIMER OF ALL WARRANTIES.
.SH BUGS
.PP
Submit bug reports online at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts/issues>>
.SH SEE ALSO
.PP
\f[B]ape2m4a\f[R](1), \f[B]ape2mp4\f[R](1), \f[B]avi2mp4\f[R](1),
\f[B]f4v2mp4\f[R](1), \f[B]flv2mp4\f[R](1), \f[B]m4a2mp3\f[R](1),
\f[B]m4v2mp4\f[R](1), \f[B]mkv2mp4\f[R](1), \f[B]mov2mp4\f[R](1),
\f[B]mp42m4v\f[R](1), \f[B]mpg2mp4\f[R](1), \f[B]png2jpg\f[R](1),
\f[B]wma2m4a\f[R](1), \f[B]wmv2m4v\f[R](1), \f[B]wmv2mp4\f[R](1)
.PP
Full documentation and sources at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts>>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Jan 27, 2022 Version 4.0.1 Release 2
    Minor release update with improved support for SD management utilities,
    extending supported platforms to Linux for the SD utils.

Jan 17, 2022 Version 4.0.1 Release 1
    Release update to Version 4.0.1 including additional man pages
    and Zip format distribution archive. Minor bug fixes and improvements.
    RPM installation fixes.

Nov 26, 2021 Version 4.0 Release 1
    This release adds Debian and RPM format installation package support with
    improved support for ease of installation, configuration, update, and removal.

Wed Jan 26 14:27:43 2022 -0800 b752940 :
   Exec /bin/bash in systemd service to comply with Fedora restrictions
Tue Jan 25 08:37:12 2022 -0800 f38080d :
   Add journalclean command
Tue Jan 25 07:14:52 2022 -0800 d4b92fc :
   Set backgrounds folder to /media/user/Transcend by default on Linux
Mon Jan 24 18:21:57 2022 -0800 363a940 :
   Change default SD device to /dev/sdd
Mon Jan 24 10:17:35 2022 -0800 3be3fab :
   Add documentation for sdmount and sdumount
Sun Jan 23 19:48:26 2022 -0800 4088f8d :
   Add sdmount and sdumount commands
Sun Jan 23 10:31:20 2022 -0800 68b537d :
   Doc for sdlist
Sun Jan 23 10:04:48 2022 -0800 04f8708 :
   Add sdlist command
Sat Jan 22 16:03:14 2022 -0800 02729e7 :
   Extend SD utilities with Linux support
Fri Jan 21 07:50:47 2022 -0800 6be5d58 :
   Muriel aliases
Thu Jan 20 11:45:13 2022 -0800 24d045d :
   Redirect stderr and stdout to parent process during RPM scriptlets
Tue Jan 18 17:17:31 2022 -0800 2152b2d :
   Use DISPLAY env var if set
Tue Jan 18 13:20:58 2022 -0800 85da437 :
   Add doctorinfo script
Mon Jan 17 13:52:51 2022 -0800 65cd49e :
   Adopt release creation during Gitlab CI
Mon Jan 17 13:42:38 2022 -0800 735ff4e :
   Do not invoke dl_fred_scripts during postinstall, new version string
Mon Jan 17 12:06:57 2022 -0800 6bc24c5 :
   Move pvoutput to MirrorCommand
Mon Jan 17 12:06:31 2022 -0800 8dd49ed :
   Bump release
Fri Jan 14 13:52:28 2022 -0800 1ab3135 :
   Update miner scripts
Fri Jan 14 10:49:45 2022 -0800 988389c :
   Add disable-update-notifier and pkg/tools/get_release
Mon Jan 10 12:23:00 2022 -0800 6e4799a :
   Do not package or export gdrive utils
Mon Jan 10 12:19:41 2022 -0800 9c85af0 :
   Remove gdrive files as they are now packaged in their own repo
Sun Jan 9 17:43:49 2022 -0800 8962417 :
   disable rpmbuild strip
Sun Jan 9 17:18:32 2022 -0800 5af1d70 :
   Fix files section of spec file
Sun Jan 9 17:10:20 2022 -0800 905d402 :
   Move all files into /usr/local
Sun Jan 9 16:39:42 2022 -0800 e27d740 :
   Try noninteractive rpm install again
Sun Jan 9 15:35:04 2022 -0800 64ddc3a :
   Fix for CI
Sun Jan 9 15:28:21 2022 -0800 fd5915b :
   export noninteractive
Sun Jan 9 15:18:00 2022 -0800 c6e51e9 :
   Copy rpm dist tree in install phase
Sun Jan 9 14:46:58 2022 -0800 bfa4e07 :
   Install rpm in CI in noninteractive mode with TZ set
Sun Jan 9 14:17:31 2022 -0800 53255f8 :
   Use ubuntu:latest CI image
Sun Jan 9 14:09:05 2022 -0800 c6963e9 :
   Version 4.0 release 3
Sun Jan 9 14:04:52 2022 -0800 fd33ea1 :
   Install zip if not found, create zip artifacts in CI
Thu Jan 6 11:57:39 2022 -0800 3b3f004 :
   Add drive to projects
Mon Jan 3 20:15:25 2022 -0800 748b3e0 :
   Add author attribution
Mon Jan 3 16:22:13 2022 -0800 236f864 :
   Add -l option to list sync folders
Mon Jan 3 16:11:26 2022 -0800 d30988a :
   Renamed getfileids as getfolderids
Mon Jan 3 13:01:47 2022 -0800 97d0e8d :
   Check for gdrive binary in PATH
Mon Jan 3 12:38:50 2022 -0800 ded8f3d :
   Add options to getfolderids, start on support for spaces in file and folder names
Mon Jan 3 10:24:36 2022 -0800 627a90d :
   Support spaces in folder names, retrieve all folder ids
Sun Jan 2 21:10:06 2022 -0800 1ec7cb9 :
   Export gdrive utils
Sun Jan 2 21:07:06 2022 -0800 40e2bdb :
   Add gdrive scripts to exports
Sun Jan 2 21:04:30 2022 -0800 e9d2a92 :
   Add comment for developer's guide to api
Sun Jan 2 20:19:56 2022 -0800 c674bc0 :
   Add sync_from_drive, update sync2drive
Sun Jan 2 11:48:46 2022 -0800 285acd8 :
   Updated gdrive utility scripts
Sat Jan 1 20:14:40 2022 -0800 a670a0d :
   Add new scripts for command line management of Google Drive
Fri Dec 31 08:52:06 2021 -0800 a1659d2 :
   Add Mirco Cabbia alias
Tue Dec 28 09:50:38 2021 -0800 904ab6d :
   Rename MirrorCommandLine to MirrorCommand
Mon Dec 27 18:20:34 2021 -0800 3554d70 :
   New model alias
Mon Dec 27 08:43:35 2021 -0800 5a2a524 :
   New model alias
Mon Dec 27 08:23:33 2021 -0800 a981875 :
   Create man page for md2man
Sun Dec 26 17:48:24 2021 -0800 6d83be7 :
   Set output dir based on man page section in md2man
Sat Dec 25 08:04:35 2021 -0800 ef79abb :
   Added Colombian wallpapers
Fri Dec 24 11:34:02 2021 -0800 0c9e9cc :
   Adjusted aliases
Thu Dec 23 09:38:25 2021 -0800 c2b5191 :
   Updated model name, add Mila Azul to Ukrainians
Wed Dec 22 16:31:35 2021 -0800 d4296b4 :
   Use HEAD to determine branch
Tue Dec 21 14:01:20 2021 -0800 526124a :
   Bump release number, update changelog
Tue Dec 21 10:26:02 2021 -0800 8c3a636 :
   Add all my clone and build scripts
Tue Dec 21 09:43:04 2021 -0800 d625a49 :
   Check and set location of Synergy client binary and system tools in Synergy startup control script
Tue Dec 21 09:14:43 2021 -0800 7a101e1 :
   Add man page for sdwrite, add SEE ALSO pages to all SD card commands
Tue Dec 21 08:40:29 2021 -0800 58b2dfc :
   Move sdwrite from MirrorCommandLine package to here
Sat Dec 18 17:18:51 2021 -0800 148f60a :
   Set synergyctl IP in postinstall
Sat Dec 18 16:39:57 2021 -0800 a8df6ce :
   Add synergyctl script to control Synergy start/stop
Fri Dec 17 07:15:45 2021 -0800 30fd5e5 :
   New model aliases
Mon Dec 13 21:11:37 2021 -0800 2c51ae3 :
   Fix typo in rmbrokenpkg
Mon Dec 13 20:18:02 2021 -0800 e5436b5 :
   Add rmbrokenpkg
Sun Dec 12 13:40:26 2021 -0800 2a4714d :
   Add note about Windows System for Linux

any2any.sh
#!/bin/bash
#
## @file any2any.sh
## @brief Convert from any audio/video format to any other
## @author Ronald Joe Record (rr at ronrecord dot com)
## @copyright Copyright (c) 2014, Ronald Joe Record, all rights reserved.
## @date Written 7-Mar-2014
## @version 1.0.1
##
##   This program works by either linking or copying any2any to a file
##   which specifies the desired input and output formats by its name.
##   Alternately, the -i and -o command line options can be used to specify
##   the input and output file formats.
##
##   For example, if you want to convert from WMV to MP4 then you could
##   create a symbolic link from any2any to wmv2mp4 as follows:
##       ln -s any2any wmv2mp4
##   Similarly, symbolic links (or copies or hard links) could be created
##   to convert from any (3 lowercase letter representation) audio/video format
##   to any other audio/video format. Commonly used conversions include:
##       wmv2mkv avi2mpg wmv2mp4 and so on.
##
##   Naming restricton: [3 lowercase letter input]2[3 lowercase letter output]
##   for a 7 letter name with "2" in the middle. The 3 letter prefix and suffix
##   must also be a filename suffix that ffmpeg recognizes as a valid
##   audio/video format.
##
##   For example, the following are filenames that conform to this restriction:
##       ape2m4a flv2mp4 mkv2mp4 wmv2mp4 avi2mpg wmv2mkv wma2m4a
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# The Software is provided "as is", without warranty of any kind, express or
# implied, including but not limited to the warranties of merchantability,
# fitness for a particular purpose and noninfringement. In no event shall the
# authors or copyright holders be liable for any claim, damages or other
# liability, whether in an action of contract, tort or otherwise, arising from,
# out of or in connection with the Software or the use or other dealings in
# the Software.
#

ME=`basename $0`
if [ "$ME" = "any2any" ]
then
    PRE="wmv"
    SUF="mp4"
    INP="WMV"
    DIR="MP4"
else
    PRE="${ME:0:3}"
    SUF="${ME:4:3}"
    INP=`echo $PRE | tr '[:lower:]' '[:upper:]'`
    DIR=`echo $SUF | tr '[:lower:]' '[:upper:]'`
fi

CODEC=
SIZE=
OVER=
TELL=
USAGE=
ACODEC=
VCODEC=
PRESET="-preset slow"
QSCALE=
RATE=
ASAMPLE=
VSAMPLE=
THREAD=
ITUNES=

## @fn usage()
## @brief Display command line usage options
## @param none
##
## Exit the program after displaying the usage message and example invocations
usage() {
  printf "\nUsage: $ME [-a audio codec] [-v video codec] [-c codec] [-p preset]"
  printf "\n\t\t[-i input format (lower case)] [-o output format (lower case)]"
  printf "\n\t\t[-q scale] [-r rate] [-b audio bitrate] [-z video bitrate]"
  printf "\n\t\t[-s size] [-t threads] [-I] [-d] [-y] [-u] file1 [file2 ...]"
  printf "\n\nWhere:\n\t-d indicates tell me what you would do"
  printf "\n\t-i input format specifies the 3 letter lower case input format"
  printf "\n\t-o output format specifies the 3 letter lower case output format"
  printf "\n\t-a audio codec specifies the output audio codec"
  printf "\n\t-v video codec specifies the output video codec"
  printf "\n\t-s size specifies the output video size (widthxheight)"
  printf "\n\t-c codec specifies the output codec"
  printf "\n\t-I indicates add the converted file to iTunes"
  printf "\n\t-p preset specifies the ffmpeg preset to use"
  printf "\n\t\t Useful presets:"
  printf "\n\t\t ultrafast superfast veryfast faster fast medium slow"
  printf "\n\t\t slower veryslow. Default preset is 'slow'"
  printf "\n\t-q scale specifies the qscale variable bit rate quality"
  printf "\n\t-r rate specifies the rate for Constant Rate Factor (CRF)"
  printf "\n\t\tencoding. Use \"-r 0\" to disable for formats other than x264"
  printf "\n\t-b bitrate specifies the bitrate (default 128k)"
  printf "\n\t-t threads specifies the number of threads to use"
  printf "\n\t-y indicates overwrite output files without asking"
  printf "\n\t-u displays this usage message\n"
  printf "\nCurrent invocation defaults to:\n\nffmpeg -i INPUT -loglevel warning -strict -2 $OPTS OUTPUT\n"
  printf "\nExamples:"
  printf "\n\tThe most common use is conversion using the default parameters"
  printf "\n\t\t$ME $INP/example.$SUF"
  printf "\n\tCopy audio and video codecs, convert the container"
  printf "\n\t\t$ME -c copy $INP/example.$SUF"
  printf "\n\tSpecify output video size of 640 width by 480 height"
  printf "\n\t\t$ME -s 640x480 $INP/example.$SUF"
  printf "\n\tSpecify libx264 output video codec and use ipod640 preset"
  printf "\n\t\t$ME -v libx264 -p ipod640 $INP/example.$SUF"
  printf "\n\tSpecify 512k bitrate and add converted file to iTunes"
  printf "\n\t\t$ME -b 512k -I $INP/example.$SUF"
  printf "\n\tSpecify libx264 output video codec and variable bit rate"
  printf "\n\t\t$ME -v libx264 -r 0 -q 3 $INP/example.$SUF\n"
  exit 1
}

while getopts i:o:p:q:r:b:t:s:v:z:a:c:Iydu flag; do
    case $flag in
        a)
            ACODEC="-acodec $OPTARG"
            ;;
        b)
            ASAMPLE="-ab $OPTARG"
            ;;
        i)
            PRE=$OPTARG
            INP=`echo $PRE | tr '[:lower:]' '[:upper:]'`
            ;;
        o)
            SUF=$OPTARG
            DIR=`echo $SUF | tr '[:lower:]' '[:upper:]'`
            ;;
        z)
            VSAMPLE="-vb $OPTARG"
            ;;
        c)
            CODEC="-codec $OPTARG"
            ;;
        d)
            TELL=1
            ;;
        i)
            inst=`type -p osascript`
            if [ "$inst" ]
            then
                ITUNES=1
            else
                echo "AppleScript is not supported on this platform."
                echo "Unable to automate the installation in iTunes."
            fi
            ;;
        p)
            PRESET="-preset $OPTARG"
            ;;
        q)
            QSCALE="-qscale $OPTARG"
            RATE=
            ;;
        r)
            if [ $OPTARG -eq 0 ]
            then
                RATE=
            else
                RATE="-crf $OPTARG"
            fi
            ;;
        s)
            SIZE="-s $OPTARG"
            ;;
        t)
            THREAD="-threads $OPTARG"
            ;;
        y)
            OVER="-y"
            ;;
        v)
            VCODEC="-vcodec $OPTARG"
            ;;
        u)
            USAGE=1
            ;;
    esac
done
shift $(( OPTIND - 1 ))

# Set ffmpeg conversion options based upon the desired output format
[ "$SUF" = "mp4" ] && {
    VCODEC="-vcodec libx264"
## See http://slhck.info/articles/crf for info on Constant Rate Factor (CRF)
    RATE="-crf 23"
    ASAMPLE="-ab 128k"
    THREAD="-threads 0"
}
[ "$SUF" = "wmv" ] && {
    ACODEC="-acodec adpcm_ms"
    VCODEC="-vcodec msmpeg4"
## See http://www.kilobitspersecond.com/2007/05/24/ffmpeg-quality-comparison
    QSCALE="-qscale 5"
    ASAMPLE="-ab 128k"
    THREAD="-threads 0"
}
[ "$SUF" = "avi" ] && {
    ACODEC="-acodec adpcm_ms"
    VCODEC="-vcodec libxvid"
    QSCALE="-qscale 5"
    ASAMPLE="-ab 128k"
    THREAD="-threads 0"
}
# TODO: Add desired default settings for other format encodings here.
#
# Set any options specified on the command line - the sed part is just
# removing any leading or trailing spaces.
OPTS=`echo "$OVER $ACODEC $CODEC $SIZE $VCODEC $PRESET $QSCALE $RATE $ASAMPLE $VSAMPLE $THREAD" | sed -e 's/^ *//' -e 's/ *$//'`
# Check for incompatible ffmpeg arguments
[ "$RATE" ] && [ "$QSCALE" ] && {
    printf "\nCannot specify both constant and variable bit rate values."
    printf "\nCurrent invocation would invoke:\n\nffmpeg -i INPUT -loglevel warning -strict -2 $OPTS OUTPUT\n"
    printf "\nExiting.\n"
    exit 1
}

# Display usage message and exit if -u was given on the command line.
# We delay until now in order to gather up all the other command line options.
[ "$USAGE" ] && usage

[ -d "$DIR" ] || {
    printf "\nOutput directory $DIR does not exist or is not a directory.\n"
    while true
    do
        read -p "Do you want to create the output directory ? (y/n) " yn
        case $yn in
            [Yy]* ) mkdir -p "$DIR"; break;;
            [Nn]* ) printf "\nExiting.\n"; exit 1;;
                * ) echo "Please answer yes or no.";;
        esac
    done
}

for i in "$@"
do
    [ -f "$i" ] || {
        echo "Input file $i does not exist or is not a regular file. Skipping."
        continue
    }
    # Set the output directory and filename
    d=`dirname "$i"`
    if [ "$d" = "." ]
    then
        j=${DIR}/`echo "$i" | sed -e "s/\.${PRE}/\.${SUF}/"`
    else
        # I always have to guess how many backslashes to use. I love sed.
        # Anyway, we need to turn slashes into backslashed slashes so the
        # subsequent sed to set the output directory will work.
        INP=`echo "$d" | sed -e 's/\//\\\\\//g'`
        j=`echo "$i" | sed -e "s/${INP}/${DIR}/" -e "s/\.${PRE}/\.${SUF}/"`
    fi
    echo "ffmpeg -i $i -loglevel warning -strict -2 $OPTS $j"
    [ "$TELL" ] || {
        ffmpeg -i "$i" -loglevel warning -strict -2 $OPTS "$j"
        [ "$ITUNES" ] && {
            # Construct full pathname of output file
            k=`pwd`/"$j"
            echo "Adding $j to iTunes"
            osascript -e "tell application \"iTunes\" to add POSIX file \"$k\""
        }
    }
done
